### PR TITLE
client: support no_pivot_root in exec driver configuration

### LINKED ISSUE #7136: Nomad exec driver does not support rootfs root partition
### Nomad version
Nomad v0.10.2

### Operating system and Environment details
Debian Stretch
Root partition is a rootfs (ramdisk)
```
$ df -h /
Filesystem      Size  Used Avail Use% Mounted on
rootfs           63G  4.5G   59G   8% /
```

### Issue
Any job using the `exec` driver fails to start which I believe because `pivot_root` does not work when the root partition is a ramdisk, see `man 2 pivot_root`
```
The rootfs (initial ramfs) cannot be pivot_root()ed.
```

I think Docker addressed this issue with the `DOCKER_RAMDISK` env variable which essentially disables `pivot_root`. I think what I need here is a way to disable `pivot_root`, probably as a configuration variable in the `exec` driver.

### Reproduction steps

1. Boot a Ubuntu live cd
1. Run Nomad
1. Plan a job using the `exec` driver

### Job file (if appropriate)
```
$ cat /etc/nomad-jobs/job-hello_world.hcl 
job "hello_world" {
  datacenters = ["foo"]

    task "server" {
      driver = "exec"

      config {
        command = "/bin/true"
      }
  }
}
```

### Nomad Client logs (if appropriate)

```
Feb 12 19:49:03 machine nomad-client[258433]:     2020-02-12T19:49:03.332Z [ERROR] client.alloc_runner.task_runner: running driver failed: alloc_id=fca82002-653b-33df-9324-4ba2bc1bbd15 task=server error="failed to launch command with executor: rpc error: code = Unknown desc = container_linux.go:346: starting container process caused "process_linux.go:449: container init caused \"rootfs_linux.go:109: jailing process inside rootfs caused \\\"pivot_root invalid argument\\\"\"""
```

### PR DESCRIPTION
**Pull request by th0m**
Adds a new `no_pivot_root` plugin option to the `exec` plugin.
It defaults to `false`, changing it to `true` will pass the `NoPivotRoot` configuration option to `libcontainer` which will fall back to using the `msMoveRoot` function for isolation.
This is useful for systems where the root is on a ramdisk.

Fixes #7136

### CODE REVIEW

File: drivers/exec/driver.go [hunk 1]
```diff
@@ -88,6 +93,9 @@ type Driver struct {
 	// event can be broadcast to all callers
 	eventer *eventer.Eventer
 
+	// config is the driver configuration set by the SetConfig RPC
+	config *Config
```
Comment (notnoop) [inline drivers/exec/driver.go]: `d.config` might be nil in tests - hence one test failure.  You may consider avoid using the pointer here.

```suggestion
	config Config
```

File: drivers/shared/executor/proto/executor.proto [hunk 2]
```diff
@@ -28,9 +28,10 @@ message LaunchRequest {
     string task_dir = 8;
     bool resource_limits = 9;
     bool basic_process_cgroup = 10;
-    repeated hashicorp.nomad.plugins.drivers.proto.Mount mounts = 11;
-    repeated hashicorp.nomad.plugins.drivers.proto.Device devices = 12;
-    hashicorp.nomad.plugins.drivers.proto.NetworkIsolationSpec network_isolation = 13;
+    bool no_pivot_root = 11;
```
Comment (notnoop) [inline drivers/shared/executor/proto/executor.proto]: I would insert append this to the end as item 14.  For backward compatibility, field numbers don't change once they are assigned.

**Comment by th0m** [comment]
Thanks for the review @notnoop. I'll add the integration test and address your comments.
This PR relates to issue #7136

**Comment by th0m** [comment]
I added a test and addressed your comments @notnoop, let me know if that looks good and thanks again for the feedback!
I still have a failed build for {{lint-go}} which I think was due to the fact I am using protobuf 3.11.3, I downgraded to 3.6.1 which is what CI seems to be using and the {{make proto}} output lined up with what was in CI now.

**Comment by th0m** [comment]
All checks are green, looks like the `protobuf` downgrade before generating proto files fixed my `lint-go` build problem.

File: website/pages/docs/drivers/exec.mdx [hunk 3]
```diff
@@ -93,6 +93,13 @@ If you are receiving the error:
 and using the exec driver, check to ensure that you are running Nomad as root.
 This also applies for running Nomad in -dev mode.
 
+## Plugin Options
+
+* `no_pivot_root` - Defaults to `false`. Changing this to `true` will pass the
+  `NoPivotRoot` configuration option to `libcontainer` which will fall back to
+  using the `msMoveRoot` function for isolation. This is useful for systems
+  where the root is on a ramdisk.
```
Comment (notnoop) [inline website/pages/docs/drivers/exec.mdx]: Some of this feels like an implementation details (e.g. the bits about libcontainer).  How do you feel about something like:

```md
* `no_pivot_root` - Defaults to `false`. When `true`, the driver uses `chroot` for file system
  isolation without `pivot_root`.  This is useful for systems where the root is on a ramdisk.
```

Comment (th0m) [reply website/pages/docs/drivers/exec.mdx]: Sure, that sounds good! I just updated the PR.

**Pull request merged**